p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,688 @@
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	host "github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	gostream "github.com/libp2p/go-libp2p/p2p/gostream"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const ProtocolIDForMultistreamSelect = "/http/1.1"
+const PeerMetadataLimit = 8 << 10 // 8KB
+const PeerMetadataLRUSize = 256   // How many different peer's metadata to keep in our LRU cache
+
+// The scheme we use internally to name a stream based roundtripper. This is only used locally.
+const libp2pStreamScheme = "libp2p-stream"
+
+// TODOs:
+// - integrate with the conn gater and resource manager
+
+type WellKnownProtocolMeta struct {
+	Path string `json:"path"`
+}
+
+type WellKnownProtoMap map[protocol.ID]WellKnownProtocolMeta
+
+// WellKnownHandler is an http.Handler that serves the .well-known/libp2p resource
+type WellKnownHandler struct {
+	wellknownMapMu   sync.Mutex
+	wellKnownMapping WellKnownProtoMap
+}
+
+// StreamHostListen retuns a net.Listener that listens on libp2p streams for HTTP/1.1 messages.
+func StreamHostListen(streamHost host.Host) (net.Listener, error) {
+	return gostream.Listen(streamHost, ProtocolIDForMultistreamSelect)
+}
+
+func (h *WellKnownHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Check if the requests accepts JSON
+	accepts := r.Header.Get("Accept")
+	if accepts != "" && !(strings.Contains(accepts, "application/json") || strings.Contains(accepts, "*/*")) {
+		http.Error(w, "Only application/json is supported", http.StatusNotAcceptable)
+	}
+
+	if r.Method != "GET" {
+		http.Error(w, "Only GET requests are supported", http.StatusMethodNotAllowed)
+	}
+
+	// Return a JSON object with the well-known protocols
+	h.wellknownMapMu.Lock()
+	mapping, err := json.Marshal(h.wellKnownMapping)
+	h.wellknownMapMu.Unlock()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Header().Add("Content-Type", "application/json")
+	w.Header().Add("Content-Length", strconv.Itoa(len(mapping)))
+	w.WriteHeader(http.StatusOK)
+	w.Write(mapping)
+}
+
+func (h *WellKnownHandler) AddProtocolMapping(p protocol.ID, path string) {
+	h.wellknownMapMu.Lock()
+	if h.wellKnownMapping == nil {
+		h.wellKnownMapping = make(map[protocol.ID]WellKnownProtocolMeta)
+	}
+	h.wellKnownMapping[p] = WellKnownProtocolMeta{Path: path}
+	h.wellknownMapMu.Unlock()
+}
+
+func (h *WellKnownHandler) RmProtocolMapping(p protocol.ID, path string) {
+	h.wellknownMapMu.Lock()
+	if h.wellKnownMapping != nil {
+		delete(h.wellKnownMapping, p)
+	}
+	h.wellknownMapMu.Unlock()
+}
+
+// HTTPHost is a libp2p host for request/responses with HTTP semantics. This is
+// in contrast to a stream-oriented host like the host.Host interface. Warning,
+// this is experimental. The API will likely change.
+type HTTPHost struct {
```

Comment:
In analogy with `net/http`, should this be called `Server`?

I'm also wondering if we should get rid of the constructor (`net/http.Server` doesn't have one either). My first attempt at setting up a server was:
```go
host := libp2phttp.HTTPHost{}
host.SetHttpHandler("/website", handler)
log.Fatal(host.Serve())
```
which panicked immediately.

Reply:
> In analogy with net/http, should this be called Server?

This can be used by the client as well (just like the stream host).

> I'm also wondering if we should get rid of the constructor

Not opposed. Should options be public fields as well then?

Reply:
I've made this change. I've removed the constructor entirely and options are set by setting the public fields.